pkg/internal/testhelpers/containers.go

```diff
@@ -130,24 +236,145 @@ func (s stdoutLogConsumer) Accept(l testcontainers.Log) {
 }
 
 // StartPGContainer starts a postgreSQL container for use in testing
-func StartPGContainer(ctx context.Context, usePromscaleExtension bool, useTimescaleDB bool, useTimescale2 bool, testDataDir string, printLogs bool) (testcontainers.Container, error) {
+func StartPGContainer(
+	ctx context.Context,
+	extensionState ExtensionState,
+	testDataDir string,
+	printLogs bool,
+) (testcontainers.Container, io.Closer, error) {
 	var image string
-	if usePromscaleExtension && useTimescale2 {
+	switch extensionState {
+	case MultinodeAndPromscale:
+		image = "timescaledev/promscale-extension:2.0.0-rc2-pg12"
+	case Multinode:
+		image = "timescale/timescaledb:2.0.0-rc2-pg12"
+	case Timescale2AndPromscale:
 		image = "timescaledev/promscale-extension:2.0.0-rc2-pg12"
-	} else if useTimescale2 {
+	case Timescale2:
 		image = "timescale/timescaledb:2.0.0-rc2-pg12"
-	} else if usePromscaleExtension && useTimescaleDB {
+	case Timescale1AndPromscale:
 		image = "timescaledev/promscale-extension:latest-pg12"
-	} else if useTimescaleDB {
+	case Timescale1:
 		image = "timescale/timescaledb:latest-pg12"
-	} else {
+	case VanillaPostgres:
 		image = "postgres:12"
 	}
-	return StartPGContainerWithImage(ctx, image, testDataDir, "", printLogs, usePromscaleExtension, useTimescaleDB)
+
+	return StartPGContainerWithImage(ctx, image, testDataDir, "", printLogs, extensionState)
 }
 
-func StartPGContainerWithImage(ctx context.Context, image string, testDataDir string, dataDir string, printLogs bool, usePromscaleExtension bool, useTimescaleDB bool) (testcontainers.Container, error) {
+func StartPGContainerWithImage(ctx context.Context,
+	image string,
+	testDataDir string,
+	dataDir string,
+	printLogs bool,
+	extensionState ExtensionState,
+) (testcontainers.Container, io.Closer, error) {
+
+	multinode = extensionState.usesMultinode()
+
+	c := CloseAll{}
+	var networks []string
+	if multinode {
+		networkName, closer, err := createMultinodeNetwork()
+		if err != nil {
+			return nil, c, err
+		}
+		c.Append(closer)
+		networks = []string{networkName}
+	}
+
+	startContainer := func(containerPort nat.Port) (testcontainers.Container, error) {
```

Comment:
This has a code smell. You are only creating this closure to avoid passing in all these parameters in multiple places. We should probably create an object (ClusterConfig?) that stores these configs and it can have StartCluster and StartInstance methods.

Reply:
I agree that this function needs to be refactored; honestly I think most of `testhelpers` needs to be (in particular the way we're storing `host` and `port` as global state has bitten us a few times). However, I think it's only as much a priority as the other refactors we have on the list, like moving `end_to_end_tests` or breaking up `sql_ingest.go`; for the same reason we decided to prioritize 2.0 support over finishing those, I think we should do the same here, and revisit this after multinode is working. (I'm not entirely sure that a config struct solves the problem here: config structs make things less fraught for callers since they give arguments names. However, here we should only ever have one caller, since this is a private helper function, and most of the arguments are internal details. I think the issue we're actually running into is that the container organization has implications to our test logic, but we don't have a good way to separate out container-starting from container-organization because testcontainers requires both of them at container start)

(I also wonder if there are ways we can refactor this to share containers across more tests, it might help with some of our CI issues).

And because I'm so easily nerd-sniped:
I'm not entirely sure that a lamda capturing the contextual arguments is bad style in the general case: it's pretty much equivalent to the partial function evaluation
```haskell
startConainer = startPGInstance ctx image testDataDir dataDir extensionState printLogs networks
```
though with a slightly uglier style since go doesn't have partial evaluation. This contains strictly more information than a config struct would; it says not only that these function calls have the same arguments, it says they _must_ have identical arguments, and only the remaining inputs can change. With a config struct it's ambiguous which of the arguments are set for convenience and which are set out of necessity.

Reply:
Let's add this to the refactorpalooza list then